File: JavaScript/1552-Magnetic Force Between Two Balls.js
/**
 * @param {number[]} position
 * @param {number} m
 * @return {number}
 */
var maxDistance = function(position, m) {
    const N = position.length;
    
    position.sort(function(a,b){
        return a-b;
    })
    
    let putBalls = (diff) =>{
        let count = 1, prev = position[0];
        for (let i=1; i<N; i++){
            if (position[i] - prev >= diff){
                count++;
                if (count === m) return true;
                prev = position[i];
            }
        }
        return false;
    }
    
    let left = 1, right = Math.floor((position[N-1] - position[0])/(m-1));
    while (left + 1 < right){
        let mid = left + Math.floor((right-left)/2);
        if (putBalls(mid)){
            left = mid;
        }else{
            right = mid;
        }  
    }
    
    if (putBalls(right)){
        return right;
    }else{
        return left;
    }
};

File: C++/1552-Magnetic Force Between Two Balls.cpp
class Solution {
public:
    bool isOK(vector<int>& position, int dist, int m) {
        int prev = position[0];
        int cnt = 1;
        for (int i=1; i<position.size(); i++) {
            if ((position[i] - prev) >= dist) {
                cnt++;
                prev = position[i];
            } 
        }
        return cnt >= m;
    }
    int maxDistance(vector<int>& position, int m) {
        sort(position.begin(), position.end());
        int left = 0, right = position.back();
        while (left + 1 < right) {
            int mid = left + (right-left)/2;
            if (isOK(position, mid, m)) {
                left = mid;
            } else {
                right = mid;
            }
        }

        if (isOK(position, right, m)) {
            return right;
        }
        return left;
    }
};